From 533c469d60cbcdac40a1b01f3ebbca9ab27fe269 Mon Sep 17 00:00:00 2001
From: Michal Schmidt <mschmidt@redhat.com>
Date: Fri, 5 Sep 2014 19:49:10 -0400
Subject: [ethernet] bnx2x: fix L2-GRE TCP issues

Message-id: <1409946615-5862-6-git-send-email-mschmidt@redhat.com>
Patchwork-id: 91551
O-Subject: [RHEL7.1 PATCH 05/70] bnx2x: fix L2-GRE TCP issues
Bugzilla: 1089402
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1089402

Upstream commit:
commit e8c37aff2a392c31fefa62ec045058a0877c45a1
Author: Dmitry Kravkov <dmitry@broadcom.com>
Date:   Tue Feb 4 17:43:03 2014 +0200

    bnx2x: fix L2-GRE TCP issues

    When configuring GRE tunnel using OVS, tcp stream is distributed over
    all RSS queues which may cause TCP reordering. It happens since OVS
    uses L2GRE protocol when kernel gre uses IPGRE.
    Patch defaults gre tunnel to L2GRE which allows proper RSS for L2GRE
    packets and (implicitly) disables RSS for IPGRE traffic.

    Signed-off-by: Dmitry Kravkov <dmitry@broadcom.com>
    Signed-off-by: Ariel Elior <ariele@broadcom.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.h b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.h
index 7ebbddc..79e4eab 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.h
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.h
@@ -935,7 +935,7 @@ static inline int bnx2x_func_start(struct bnx2x *bp)
  else /* CHIP_IS_E1X */
   start_params->network_cos_mode = FW_WRR;
 
- start_params->gre_tunnel_mode = IPGRE_TUNNEL;
+ start_params->gre_tunnel_mode = L2GRE_TUNNEL;
  start_params->gre_tunnel_rss = GRE_INNER_HEADERS_RSS;
 
  return bnx2x_func_state_change(bp, &func_params);
-- 
1.7.1